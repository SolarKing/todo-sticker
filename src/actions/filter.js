// @flow

export type FilterActions =
    | DueDateFilterAction;

export type DueDateFilterAction = {|
    type: 'dueDateFilter',
    filter: string,
|}

export function settingDueDateFilter(dueDateFilter: string): DueDateFilterAction {
    return {
        type: 'dueDateFilter',
        filter: dueDateFilter
    };
}
