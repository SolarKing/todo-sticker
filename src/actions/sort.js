// @flow
import type {SortByType, SortColumn} from '../constant/sort';

export type SortActions =
    | SortByAction
    | SortColumnAction;

export type SortColumnAction = {|
    type: 'sortColumn',
    column: SortColumn
|};

export type SortByAction = {|
    type: 'sortBy',
    sortBy: SortByType
|};

export function settingSortColumn(column: SortColumn): SortColumnAction {
    return {
        type: 'sortColumn',
        column: column
    };
}
