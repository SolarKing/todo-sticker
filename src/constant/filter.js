// @flow
export type DueDateFilter = 'today' | 'all';

export const DueDateFilterEnum = {
    today: 'today',
    all: 'all'
};
