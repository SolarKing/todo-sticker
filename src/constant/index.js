// @flow

const DueDate = {
    today: 'today',
    all: 'all'
};


export const FilterEnum = {
    DueDate
};
