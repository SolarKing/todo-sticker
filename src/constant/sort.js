// @flow
export type SortColumn = 'dueDate' | 'createTime';
export type SortByType = 'desc' | 'asc';

export const SortColumnEnum = {
    dueDate: 'dueDate',
    createTime: 'createTime'
};

export const SortByTypeEnum = {
    desc: 'desc',
    asc: 'asc'
};
