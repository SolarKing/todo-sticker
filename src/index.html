<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>Todo Sticker</title>
	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
</head>
<body>
	<div id="todoModal"></div>
	<div id="app"></div>
</body>
</html>
