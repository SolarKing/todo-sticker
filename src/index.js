// @flow

import React from 'react';
import ReactDOM from 'react-dom';
import TodosPage from './component/todosPage';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from './reducers';
import '../node_modules/antd/dist/antd.min.css';
import '../node_modules/antd/dist/antd.min.js';
import type {TodoState} from './states/index';
import NotifyServer from './class/notify/notifyServer';

// localStorage.removeItem('todos');
const initStorageData = localStorage.getItem('todos') || '{}';
const store = createStore(rootReducer, JSON.parse(initStorageData));
store.subscribe(() => {
    localStorage.setItem('todos', JSON.stringify(store.getState()));
    console.log(store.getState());
});

let todayTimestamp: number = new Date().getTime();

store.getState().todos.forEach((todo: TodoState) => {
    if (todo.complete) {
        return;
    }
    if (todo.dueDatetime <= todayTimestamp) {
        return;
    }
    NotifyServer.instance.addSchedule(todo.id || 'empty', todo.dueDatetime, todo.description);
});

const app = document.getElementById('app');

app && ReactDOM.render(
    <Provider store={store}>
        <TodosPage />
    </Provider>,
    app
);
