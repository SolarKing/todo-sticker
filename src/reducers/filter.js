// @flow
import type {FilterState} from '../states/index';
import {DueDateFilterEnum} from '../constant/filter';
import type {FilterActions} from '../actions/filter';

const initialFilter: FilterState = {
    dueDate: DueDateFilterEnum.all,
};

function filter(state: FilterState = initialFilter, action: FilterActions) {
    switch (action.type) {
    case 'dueDateFilter':
        return {
            ...state,
            dueDate: action.filter
        };
    default:
        return  state;
    }
}

export default filter;
