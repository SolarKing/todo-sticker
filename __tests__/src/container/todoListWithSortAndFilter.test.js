// @flow
import React from 'react';
import { shallow, type ShallowWrapper } from 'enzyme';
import TodoListWithSortAndFilter, {
    filterWithDueDate,
    filterWithCompleteStatus,
    sortByColumn
} from '../../../src/container/todoListWithSortAndFilter';
import configureStore from 'redux-mock-store';
import { type AppState } from '../../../src/states';
import { DueDateFilterEnum, CompleteStatusFilterEnum } from '../../../src/constant/filter';
import { SortByTypeEnum, SortColumnEnum } from '../../../src/constant/sort';

describe('TodoListWithSortAndFilter', () => {
    let wrapper: ShallowWrapper;
    let todos = [{
        id: '0',
        dueDatetime: new Date(2019, 1, 15).getTime(),
        description: 'Hello',
        complete: false,
        createTime: new Date(2019, 1, 10).getTime(),
    },{
        id: '1',
        dueDatetime: new Date(2019, 1, 18).getTime(),
        description: 'Hello',
        complete: true,
        createTime: new Date(2019, 1, 11).getTime(),
    },{
        id: '2',
        dueDatetime: new Date(2019, 1, 10).getTime(),
        description: 'Hello',
        complete: true,
        createTime: new Date(2019, 1, 12).getTime(),
    },{
        id: '3',
        dueDatetime: new Date(2019, 3, 15).getTime(),
        description: 'Hello',
        complete: false,
        createTime: new Date(2019, 1, 14).getTime(),
    },{
        id: '4',
        dueDatetime: new Date(2019, 1, 15).getTime(),
        description: 'Hello',
        complete: true,
        createTime: new Date(2019, 1, 13).getTime(),
    }];

    const setup = (initialState: AppState) => {
        let store = configureStore()(initialState);
        wrapper = shallow(<TodoListWithSortAndFilter store={store} />);
        return {
            store,
            wrapper
        };
    };

    it('should render correct', () => {
        const { wrapper } = setup({
            todos,
            filter: {
                dueDate: DueDateFilterEnum.all,
                completeStatus: CompleteStatusFilterEnum.complete,
            },
            sort: {
                sortBy: SortByTypeEnum.desc,
                sortColumn: SortColumnEnum.createTime,
            }
        });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.prop('todos')).toEqual([
            todos[4],
            todos[2],
            todos[1],
        ]);
    });

    it('should filter duedate is today', () => {
        jest.spyOn(Date, 'now').mockReturnValue(new Date(2019, 1, 15).getTime());

        let result = filterWithDueDate(todos, DueDateFilterEnum.today);
        expect(result).toEqual([
            todos[0],
            todos[4],
        ]);
    });

    it('should not filter duedate', () => {
        jest.spyOn(Date, 'now').mockReturnValue(new Date(2019, 1, 15).getTime());

        let result = filterWithDueDate(todos, DueDateFilterEnum.all);
        expect(result).toEqual(todos);
    });

    it('should filter with complete status', () => {
        let result = filterWithCompleteStatus(todos, CompleteStatusFilterEnum.complete);
        expect(result).toEqual([
            todos[1],
            todos[2],
            todos[4],
        ]);
    });

    it('should filter with uncomplete status', () => {
        let result = filterWithCompleteStatus(todos, CompleteStatusFilterEnum.uncomplete);
        expect(result).toEqual([
            todos[0],
            todos[3],
        ]);
    });

    it('should filter with all status', () => {
        let result = filterWithCompleteStatus(todos, CompleteStatusFilterEnum.all);
        expect(result).toEqual(todos);
    });

    it('should sort asc by createTime ', () => {
        let result = sortByColumn(todos, {
            sortBy: SortByTypeEnum.asc,
            sortColumn: SortColumnEnum.createTime,
        });
        expect(result).toEqual([
            todos[0],
            todos[1],
            todos[2],
            todos[4],
            todos[3],
        ]);
    });

    it('should sort desc by createTime ', () => {
        let result = sortByColumn(todos, {
            sortBy: SortByTypeEnum.desc,
            sortColumn: SortColumnEnum.createTime,
        });
        expect(result).toEqual([
            todos[3],
            todos[4],
            todos[2],
            todos[1],
            todos[0],
        ]);
    });

    it('should sort asc by dueDate ', () => {
        let result = sortByColumn(todos, {
            sortBy: SortByTypeEnum.asc,
            sortColumn: SortColumnEnum.dueDate,
        });
        expect(result).toEqual([
            todos[2],
            todos[0],
            todos[4],
            todos[1],
            todos[3],
        ]);
    });

    it('should sort desc by dueDate ', () => {
        let result = sortByColumn(todos, {
            sortBy: SortByTypeEnum.desc,
            sortColumn: SortColumnEnum.dueDate,
        });
        expect(result).toEqual([
            todos[3],
            todos[1],
            todos[4],
            todos[0],
            todos[2],
        ]);
    });

});