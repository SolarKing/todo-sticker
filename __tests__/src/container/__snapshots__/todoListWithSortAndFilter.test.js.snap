// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListWithSortAndFilter should render correct 1`] = `
<TodoList
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  todos={
    Array [
      Object {
        "complete": true,
        "createTime": 1549987200000,
        "description": "Hello",
        "dueDatetime": 1550160000000,
        "id": "4",
      },
      Object {
        "complete": true,
        "createTime": 1549900800000,
        "description": "Hello",
        "dueDatetime": 1549728000000,
        "id": "2",
      },
      Object {
        "complete": true,
        "createTime": 1549814400000,
        "description": "Hello",
        "dueDatetime": 1550419200000,
        "id": "1",
      },
    ]
  }
/>
`;
